src/job_metadata.c

```diff
@@ -650,3 +700,238 @@ PgCronHasBeenLoaded(void)
+	else
+	{
+		values[Anum_job_run_details_return_message - 1] = CStringGetTextDatum(return_message);
+		repl[Anum_job_run_details_return_message - 1] = true;
+	}
+
+	if (start_time == NULL)
+		isNulls[Anum_job_run_details_start_time - 1] = true;
+	else
+	{
+		values[Anum_job_run_details_start_time - 1] = TimestampTzGetDatum(*start_time);
+		repl[Anum_job_run_details_start_time - 1] = true;
+	}
+
+	if (end_time == NULL)
+		isNulls[Anum_job_run_details_end_time - 1] = true;
+	else
+	{
+		values[Anum_job_run_details_end_time - 1] = TimestampTzGetDatum(*end_time);
+		repl[Anum_job_run_details_end_time - 1] = true;
+	}
+
+	schemaId = get_namespace_oid(CRON_SCHEMA_NAME, false);
+	indexId = get_relname_relid(JOB_RUN_DETAILS_INDEX_NAME, schemaId);
+
+	jobRunDetailTable = table_open(JobRunDetailsRelationId(), RowExclusiveLock);
+
+	ScanKeyInit(&scanKey[0], Anum_job_run_details_runid,
+							BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(runId));
+
+	scanDescriptor = systable_beginscan(jobRunDetailTable, indexId, true,
+									NULL, 1, scanKey);
+
+	tupleDescriptor = RelationGetDescr(jobRunDetailTable);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	if (!HeapTupleIsValid(heapTuple))
+	{
+		heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+		#if PG_VERSION_NUM < 100000
+			simple_heap_insert(jobRunDetailTable, heapTuple);
+			CatalogUpdateIndexes(jobRunDetailTable, heapTuple);
+		#else
+			CatalogTupleInsert(jobRunDetailTable, heapTuple);
+		#endif
+		heap_freetuple(heapTuple);
+	}
+	else
+	{
+		HeapTuple newtuple;
+
+		newtuple = heap_modify_tuple(heapTuple, tupleDescriptor, values, isNulls, repl);
+		#if PG_VERSION_NUM < 100000
+			simple_heap_update(jobRunDetailTable, &newtuple->t_self, newtuple);
+			CatalogUpdateIndexes(jobRunDetailTable, newtuple);
+		#else
+			CatalogTupleUpdate(jobRunDetailTable, &newtuple->t_self, newtuple);
+		#endif
+		heap_freetuple(newtuple);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(jobRunDetailTable, NoLock);
+
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+
+	return true;
+}
+
+void
+CleanAuditTable(void)
+{
+	Relation jobRunDetailTable;
+	SysScanDesc scanDescriptor;
+	HeapTuple heapTuple;
+	TupleDesc tupleDescriptor;
+	Datum values[Natts_job_run_details];
+	bool isNulls[Natts_job_run_details];
+	bool repl[Natts_job_run_details];
+
+	SetCurrentStatementStartTimestamp();
+	StartTransactionCommand();
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	/*
+	 * If the pg_job_scheduler extension has not been created yet or
+	 * we are on a hot standby, the job table is treated as
+	 * being empty.
+	 */
+	if (!PgCronHasBeenLoaded() || RecoveryInProgress())
+	{
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+		pgstat_report_activity(STATE_IDLE, NULL);
+		return;
+	}
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+	memset(repl, false, sizeof(repl));
+
+	values[Anum_job_run_details_status - 1] = CStringGetTextDatum("failed");
```

Comment:
would be nicer to use constants for these strings since they are used in a couple of places and "failed" is quite a generic string to look for.